### PR TITLE
Fixes #31080 - Add UI for managing Personal Access Tokens

### PR DESCRIPTION
**Pull request by kamils-iRonin**
This PR is an attempt to finish what's started here https://github.com/theforeman/foreman/pull/4597

image
image

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm/PersonalAccessTokenForm.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm/PersonalAccessTokenForm.js]: Is there a way to give this field a default value?  It's annoying to have to specify a time of day when you may be entering a date years in the future, so it would be nice if it would just fill in a time for you.

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokens.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokens.js]: Even though it may avoid the eslint warning, I think this use of `useEffect` still breaks [this rule](https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level) because of the `useCallback` inside of `getPersonalAccessTokens`.

I'm not sure why `useCallback` would be needed here.  If you just need to ensure that the API call only occurs on mount, you can pass an empty dependency array:
```suggestion
  useEffect(() => {
    getPersonalAccessTokensAction({ url });
  }, []);
```

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensList/PersonalAccessToken.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensList/PersonalAccessToken.js]: If I revoke a token before its expiration date, I get a value like "Expired" / "in 2 years".   For revoked tokens, would make more sense just to say "Revoked" in this cell, instead of the expiration date.

**Review by jeremylenz — COMMENTED** [review]
Thanks @kamils-iRonin !  This will be a great addition.

I didn't see any code that would actually make API requests work with the token.. is that done separately?  So I didn't try actually using a token.  But I did create and revoke some and the UI seems to work fine.

Would also like to see some documentation added on how to use tokens in an API request.

Some comments below

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm/PersonalAccessTokenForm.js
Comment (timogoebel) [reply webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm/PersonalAccessTokenForm.js]: There may be personal access tokens that never expire. How could we find a default for that?
We could provide quick actions like `in a month`, `in a year`, ... - but I believe this is premature optimization.

**Comment by ares** [comment]
@kamils-iRonin this is fantastic work a long time missing UI part.

Based on the screenshots, I have 2 questions. Does this use the same calendar component, that we already have? It looks a bit different.

Second - does it use the same `ctrl+c` code that we already have in core (for register templates)? Again, looks a bit different.

**Comment by kamils-iRonin** [comment]
> I didn't see any code that would actually make API requests work with the token.. is that done separately? So I didn't try actually using a token. But I did create and revoke some and the UI seems to work fine.
> 
> Would also like to see some documentation added on how to use tokens in an API request.

It's just an UI to manage personal access tokens. Once a token is generated, it can be used for any API request, I added an example on how to use token with curl to personal access tokens page. 

> Does this use the same calendar component, that we already have? It looks a bit different.

I simply used `TextField` component and passed `datetime-local` as type. It seems to work well with `ForemanForm`. I have not created any new calendar component.



> does it use the same `ctrl+c` code that we already have in core (for register templates)?

I used [`ClipboardCopy`](https://www.patternfly.org/v4/documentation/react/components/clipboardcopy)

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm/PersonalAccessTokenForm.js
Comment (kamils-iRonin) [reply webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm/PersonalAccessTokenForm.js]: I believe it should be empty by default (for tokens that never expire)

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm/index.js]: can we use `useDispatch()` instead of ` connect` HOC?

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokens.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokens.js]: why using  empty`[]` ? it makes the `useEffect` not being affected by `url` prop changes.
```suggestion
  useEffect(() => {
    getPersonalAccessTokens({ url });
  }, [url]);
```

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/index.js]: We added redux hooks which replace the need for  `connect` HOC, I personally find hooks much more readable and flow.
For dispatching an action  you can use `useDispatch`
For getting data from the redux' store you can use `useSelectors`

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensReducer.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensReducer.js]: any data manipulation should be done through selectors.
selectors can be memorized while reducers should act fast

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensReducer.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensReducer.js:12]: You can use our API middleware which fetches API requests and store their values in a dedicated slot in the store.
 ```js
// dispatching
    dispatch(
      get({
        key: KEY,
        url: foremanUrl(`/path`),
      })
    );
// selecting
  const yourData = useSelector(state =>
    selectAPIResponse(state, KEY) // selects the data directly from the store without extra reducer
  );
```

**Review by amirfefer — DISMISSED** [review]
Thanks @kamils-iRonin, awsome work! :+1: 
I left a few inline comments

**Comment by ares** [comment]
@kamils-iRonin what's the status on this PR? Can we help somehow to update the code? This is a great addition to the project. Thanks for letting us know.

**Comment by kamils-iRonin** [comment]
@ares I will try to find some time to apply the improvements mentioned by @amirfefer soon

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokens.js
Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokens.js]: @amirfefer It's probably from my [comment above](https://github.com/theforeman/foreman/pull/7806#discussion_r452321010), sorry :smile: 

I agree with adding `url` to the dependency array

**Comment by kamils-iRonin** [comment]
@ares I think it's ready now

**Comment by timogoebel** [comment]
@kamils-iRonin: Do you mind rebasing the PR?

**Comment by timogoebel** [comment]
@amirfefer: Were your comments addressed?

**Comment by timogoebel** [comment]
@amirfefer, @ares, @sharvit: Anything to add here?

**Comment by ares** [comment]
I'm waiting on code review from @sharvit or @amirfefer, then I'll perform testing.

File: app/views/personal_access_tokens/_personal_access_tokens_tab.html.erb
Comment (ezr-ondrej) [inline app/views/personal_access_tokens/_personal_access_tokens_tab.html.erb]: We can simplify by using the new mounter - the old one is being deprecated.
```suggestion
<%= react_component('PersonalAccessTokens', {
  url: api_user_personal_access_tokens_path(:user_id => @user),
  canCreate: authorized_for(hash_for_api_user_personal_access_tokens_path(:user_id => @user))
}) %>
```

**Review by ezr-ondrej — CHANGES_REQUESTED** [review]
Apart of the usage of the old mounter, I think this is quite OK.
What I mean is: This is not using our `Table` component and new stack with it (`Pagination`, new fetching), but given there is no documentation for it and examples in code are not of the best quality eighter, it can be quite hard to unpack, so I'd not push the contributor to do it. I'd be ok here with creating a ticket to switch this to the Table component and merge AS-IS, because what we have here is older style, but working and clean code :shrug:

**Comment by jturel** [comment]
@kamils-iRonin thank you for the contribution! I had an observation or two that you might clarify for me:

1. The Redmine issue associated is this one: https://projects.theforeman.org/issues/1301 which is more of a "tracker" with some related sub-tasks
2. This PR appears to be concerned with providing a UI for managing Personal Access Tokens (awesome)
3. The PR title is 'User login with access token' which conflicts with 2 and thereby 1

If my understanding above is correct, I'd like to ask that you:

- Create a new Redmine issue which is linked to 1301 as a sub-task
- Ensure name of the issue and this PR+commit message title reflect the intent clearly: adding UI capability around personal access tokens vs user login with access token

All of that can be done using this current PR (no need to close and open new one).

File: app/views/personal_access_tokens/_personal_access_tokens_tab.html.erb
Comment (ezr-ondrej) [inline app/views/personal_access_tokens/_personal_access_tokens_tab.html.erb]:
```suggestion
<div class="tab-pane" id="personal_access_tokes">
    <%= react_component('PersonalAccessTokens', {
    url: api_user_personal_access_tokens_path(user_id: @user),
    canCreate: authorized_for(hash_for_api_user_personal_access_tokens_path(user_id: @user))
  }) %>
</div>
```

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensList/index.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensList/index.js]:
```suggestion
  <React.Fragment>
```

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensList/index.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensList/index.js]:
```suggestion
  </React.Fragment>
```

**Review by ezr-ondrej — CHANGES_REQUESTED** [review]
The new mounter mounts component in place and wraps it in a special tag, so we need to move the pane out of the React component.
I'd argue it's a cleaner solution anyway as jquery now won't touch (hide/show on tab switch) React dom.

**Comment by jturel** [comment]
@kamils-iRonin thanks for doing the Redmine shuffle!

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensList/index.js
Comment (kamils-iRonin) [reply webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokensList/index.js]: I think `div` with `panel` class is necessary here

**Review by ezr-ondrej — DISMISSED** [review]
As stated above, this has some technical debt on the React code side, but it's a clean code quite easy to refactor once we set the clear styles for tables, thus I'd argue for getting it in AS-IS.
So I'm good here! Thanks @kamils-iRonin, great addition ! :+1:

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm.js]: Oh, I did not noticed this. Why we are not using our dateTime picker?
This won't work on browsers that doesn't support `datetime-local` field.
We can use our component from `'../../common/form/DateTimePicker'`
```suggestion
          <DateTimePicker
            name="expires_at"
            required={false}
            label={__('Expires')}
            inputClassName="col-md-6"
          />
```

**Review by ezr-ondrej — COMMENTED** [review]
Just found out, you're not using the `DateTimePicker` component here.
Is there a reason to it?

File: app/models/personal_access_token.rb
Comment (ares) [inline app/models/personal_access_token.rb]: why do we want to skip validations here?

File: app/views/users/_form.html.erb
Comment (ares) [inline app/views/users/_form.html.erb]: this only checks for user having a permission for viewing at least one user, it does not verify the permission for $this user, you need to use the following I believe (untested)
`authorized_for(:auth_action => 'view', :controller => 'personal_access_tokens', :auth_object => @user)`

File: app/views/users/_form.html.erb
Comment (ares) [inline app/views/users/_form.html.erb]: the same as above

**Review by ares — CHANGES_REQUESTED** [review]
Thanks for your great work so far. I have few suggestions from user experience point of view. I'll try to clarify it in this message, please reach out to me on IRC (mhulan) if you want have any questions.

1) The tab layout should be IMHO more consistent with SSH keys tab, compare the two screenshots
SSH tab
ssh1

PAT tab
pat1

The SSH key has some welcome text with an icon explaining the purpose, but not showing table until there's at least one entry.

ssh2

Once there is an entry, the help text is no longer displayed. We use this pattern everywhere and I think we should use it here too. If there's already one record, user probably don't need to see the help text anymore.

2) The calendar looks differently than elsewhere, can we use the calendar we already use in app?

cal1
vs input fields
cal2

3) the copy to clipboard experience is also inconsistent with other place we have, can we make it look similar or even use the same implementation? (well, I guess it's jquery vs react, so I'm fine with just css it to look more similarly)

cc1

this is from registration tab on host detail page (when host OS has registration template assigned)

4) the green button should be under tables, once there's some record in Active Personal Access Tokens table.

5) I would not put Active Personal Access Tokens and Inactive Personal Access Tokens tabels inside panes, it's third frame already (it's tab, then it's in pane, then the table border). Just keep it as h3 headers to separate those two tables.

6) there's a authorization concern I've left as a code comment

**Summary:**
I understand the design can be very subjective so I tried to keep it just to make it more consistent with what we have already in the application.
cc2

**Comment by kamils-iRonin** [comment]
I think I've applied all suggestions

- layout when there are no tokens (I'm not sure if this icon is the right one)

image

- use the `DateTimePicker` component

image

- new `ClipboardCopy` component

image

File: app/models/personal_access_token.rb
Comment (kamils-iRonin) [reply app/models/personal_access_token.rb]: there was a problem with permissions, I changed the code to
```ruby
User.as_anonymous_admin do
  token.update(last_used_at: Time.current.utc)
end
```

**Comment by ezr-ondrej** [comment]
I think, the failing snapshots have been fixed in `develop` branch. Could you rebase on top of the newest develop?
**EDIT:** It seems you're updating them yourself (probably because of old npm packages you've got older version of the snapshots). So just checking out the changed snapshots from your branch should be enough (or `npm i`, get coffee while it updates npm dependencies, update snapshots again) :)

File: webpack/assets/javascripts/react_app/components/common/DateTimePicker/DateTimePicker.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/common/DateTimePicker/DateTimePicker.js:115]: We need to wrap call to anonymous function, like this we would pass around a pointer, that could change - I don't understand exacly why, but this is not safe by React guides ^_^
```suggestion
              <Icon type="fa" name="close" onClick={() => this.clearSelected} />
```

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokenForm.js]: **OPTIONAL** The `DateTime` component is not really used and obsolate, we should use [FormField](https://theforeman.github.io/foreman/?path=/story/components-form--form-field)
```suggestion
          <FormField
            type="dateTime"
            id={UUID()}
            label={__('Expires')}
            isRequired={false}
            inputClassName="col-md-6"
            placement="bottom"
            inputProps={{ name: 'expires_at' }}
            hideValue
          />
```
Although I take as my personal failure, that this way doesn't have much of a documentation and is still not as easy to use with `Formik`, so I'm OK with merging this as is and I'll update in follow-up PR.

**Review by ezr-ondrej — COMMENTED** [review]
Needs to check out the `gridBreakpoint` snapshots, but apart of it the React side looks ok :+1:
NOTE: unless we want to backport this! Because we use the `DateTime` component in the template render invocation form now (but we will stop using it in 2.4).

File: webpack/assets/javascripts/react_app/components/common/DateTimePicker/DateTimePicker.js
Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/common/DateTimePicker/DateTimePicker.js:115]: I don't think this is necessary.. Since `<Icon>` is getting a new anonymous function as a prop, wouldn't it cause a rerender of `<Icon>` every time?  Since the function is anonymous, React is unable to compare to the previous prop to see if it changed and so will just rerender no matter what.  Rather, if left as is, `this.clearSelected` won't have changed and so may avoid rerender of `<Icon>` every time.

**Comment by ezr-ondrej** [comment]
#8122 has been merged and it introduced some changes to the DateTime component, could you please rebase, it should not be hard rebase.

**Comment by jeremylenz** [comment]
[test katello]

**Comment by ezr-ondrej** [comment]
[test katello]

**Review by ezr-ondrej — COMMENTED** [review]
Any last comments, Katello tests not rellated, no idea what's up with the Katello js tests tho

**Comment by jeremylenz** [comment]
[test katello]

is my only other comment

**Comment by ares** [comment]
Sorry, my webpack setup got broken, but is restored now, I'll retest this soon, the screens look really nice!

**Review by ares — APPROVED** [review]
Works nicely, thanks for all adjustments! This now feel much more native :-)

**Comment by ares** [comment]
I fail to see how this could cause `(non-blocking) React on plugins / test (katello, Katello) (pull_request)` error, since @jeremylenz also acked this and it's non-blocking check, I'll proceed with merging.

Thanks everyone involved in this, but mainly @kamils-iRonin. Would you be willing to demo this feature on our community demo? (next one is this thursday)

**Comment by kamils-iRonin** [comment]
> Would you be willing to demo this feature on our community demo? (next one is this thursday)

@ares sure, I can show how to create Personal Access Tokens and how to use them

**Comment by ares** [comment]
Awesome! Could you please put your name to https://community.theforeman.org/t/foreman-community-demo-88/21501 so you'll receive the presenter link before the demo?

**Comment by jeremylenz** [comment]
seems like this is causing the same test failures on katello master.

**Comment by jeremylenz** [comment]
see https://github.com/theforeman/foreman/pull/8192

File: webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokens.js
Comment (ekohl) [inline webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokens.js:70]: I just noticed this as a user, but this should use the `foreman_url` setting so I opened https://projects.theforeman.org/issues/32373.

**Pull request merged**
